platform/darwin/src/NSPredicate+MGLAdditions.mm

```diff
@@ -259,28 +271,33 @@ + (instancetype)mgl_predicateWithJSONObject:(id)object {
     NSString *op = objects.firstObject;
     
     if ([op isEqualToString:@"=="]) {
-        NSArray *subexpressions = MGLSubexpressionsWithJSONObjects([objects subarrayWithRange:NSMakeRange(1, objects.count - 1)]);
+        NSArray *sanitizedObjects = MGLSanitizeJSONObjects([objects subarrayWithRange:NSMakeRange(1, objects.count - 1)]);
```

Comment:
I removed the code you mentioned in https://github.com/mapbox/mapbox-gl-native/pull/11587#discussion_r180200037, but this code is required for comparison predicates to work as filters. `MGLSanitizeJSONObjects()` removes the data types before converting them.

Reply:
Hmm.  I think a problem here is that this is unconditionally removing all `"number"`, `"string"`, `"boolean"` assertions.  Removing the ones that were automatically inferred is fine, because they can be inferred again if the expression is sent back into mbgl to be parsed.  But it's possible to have explicitly-added assertions that would _not_ be automatically inferred if they were stripped.  For example, `["==", ["number", ["id"]], ["get", "x"]]` is an expression that specifically asserts that a feature's `id` is a number and compares it to property with key `"x"`.  If the `"number"` wrapper were stripped here, the resulting expression would fail to parse. (And even if we relaxed the typing requirements so that it _did_ parse, its behavior would be different than that of the original.)

Reply:
This applies only to `>, >=, <, <=` this type of evaluation is not possible using `NSExpression` directly.

Reply:
Ah, okay.  Some expressions might still fail to roundtrip -- e.g., `[">", ["string", ["get", "x"]], ["string", ["get", "y"]]]` would result in `[">", ["get", "x"], ["get", "y"]]` -- but I suppose that's a known/acceptable issue?

Reply:
@anandthakker @1ec5 seems like the best solution is require that the developers add `CAST` to filter predicates. Something like:
`[NSPredicate predicateWithFormat:@"CAST(x, 'NSNumber') < 5"];`
This way we avoid all the problems mentioned above.

Reply:
The code here converts from JSON to NSPredicate. As long as an expression in JSON already has type assertions, those type assertions should be reflected in the NSPredicate. The fact that this prevents literal round-tripping of some predicates is unfortunate but not a big deal compared to potentially preventing an expression from working correctly.